### PR TITLE
[Powershell/en] Refactoring Powershell Wiki

### PR DESCRIPTION
**Pull request by AndrewDavis1191**
Changing to reflect current capabilities and draw similarities to other languages in terms of capabilities

- [x] I solemnly swear that this is all original content of which I am the original author
- [x] Pull request title is prepended with `[language/lang-code]`
- [x] Pull request touches only one file (or a set of logically related files with similar changes made)
- [x] Content changes are aimed at *intermediate to experienced programmers* (this is a poor format for explaining fundamental programming concepts)
- [x] If you've changed any part of the YAML Frontmatter, make sure it is formatted according to [CONTRIBUTING.md](https://github.com/adambard/learnxinyminutes-docs/blob/master/CONTRIBUTING.markdown)
  - [x] Yes, I have double-checked quotes and field names!

### CODE REVIEW

**Comment by AndrewDavis1191** [comment]
@divayprakash are you able to review this? :innocent:

**Comment by divayprakash** [comment]
@AndrewDavis1191 Sorry I don't knwo any Powershell so won't be able to review such a large change.

**Comment by imba-tjd** [comment]
Is this expected?

图片

**Comment by AndrewDavis1191** [comment]
> ~Is this expected?~ I guess not. There are two `1. Primitive Datatypes and Operators`

Uh oh, good catch. I totally missed that. Removed in e2b7568

File: powershell.html.markdown [hunk 1]
```diff
@@ -13,116 +14,348 @@ Nearly all examples below can be a part of a shell script or executed directly
...
+-5 -ne $False # => True
+
+# Using boolean logical operators on ints casts them to booleans for evaluation
+# but their non-cast value is returned
+# Don't mix up with bool(ints) and bitwise and/or (&,|)
+[bool](0)     # => False
+[bool](4)     # => True
+[bool](-6)    # => True
+0 -and 2     # => 0
+-5 -or 0     # => -5
```
Comment (imba-tjd) [inline powershell.html.markdown:97]: My output is different

File: powershell.html.markdown [hunk 2]
```diff
@@ -13,116 +14,348 @@ Nearly all examples below can be a part of a shell script or executed directly
...
+"Hello world!"[0]  # => 'H'
+
+# You can find the length of a string
+("This is a string").Length  # => 16
+
+# You can also format using f-strings or formatted string literals
+$name = "Steve"
+$age = 22
+"He said his name is $name." # => "He said his name is Steve"
+“{0} said he is {1} years old.” -f $name, $age # => "Steve said he is 22 years old"
```
Comment (imba-tjd) [inline powershell.html.markdown:146]:
```suggestion
"{0} said he is {1} years old." -f $name, $age # => "Steve said he is 22 years old"
```

File: powershell.html.markdown (hunk 1 above)
Comment (AndrewDavis1191) [reply powershell.html.markdown]: Needed to change to -band/-bor for posh. Changed in 998249d

File: powershell.html.markdown [hunk 3]
```diff
@@ -13,116 +14,349 @@ Nearly all examples below can be a part of a shell script or executed directly
...
+# $null, 0, and empty strings and arrays all evaluate to False.
+# All other values are True
+function test ($value) {
+  if ($value) {Write-Output 'True'}
+    else {Write-Output 'False'}
+}
+test ($null) # => False
+test (0)   # => False
+test ("")  # => False
+test []  # => True
```
Comment (imba-tjd) [inline powershell.html.markdown:161]: I think this is confusing.

**Comment by imba-tjd** [comment]
```
  { 'x' -contains 'x'}    { "FALSE! -contains is for lists!"; break }
```

`'x' -contains 'x'` returns True.

File: powershell.html.markdown (hunk 3 above)
Comment (AndrewDavis1191) [reply powershell.html.markdown]: I've added some comments and details to the objects.

**Comment by AndrewDavis1191** [comment]
> ```
>   { 'x' -contains 'x'}    { "FALSE! -contains is for lists!"; break }
> ```
> 
> `'x' -contains 'x'` returns True.

This was from the original creator. It may have been true at that time, but it seems it is not true in the current version of Powershell.

**Comment by imba-tjd** [comment]
```
function test ($value) { $value.gettype() }
test []
```

The result is string

File: powershell.html.markdown [hunk 4]
```diff
@@ -13,116 +14,373 @@ Nearly all examples below can be a part of a shell script or executed directly
...
+"{0} said he is {1} years old." -f $name, $age # => "Steve said he is 22 years old"
+"$name's name is $($name.Length) characters long." # => "Steve's name is 5 characters long."
+
+# $null is not an object
+$null  # => None
+
+# $null, 0, and empty strings and arrays all evaluate to False.
+# All other values are True
+# $null, 0, and empty strings and arrays all evaluate to False.
+# All other values are True
```
Comment (imba-tjd) [inline powershell.html.markdown:155]: You know

File: powershell.html.markdown [hunk 5]
```diff
@@ -13,116 +14,373 @@ Nearly all examples below can be a part of a shell script or executed directly
...
+You cannot call a method on a null-valued expression.
+At line:4 char:1
++ $test3.getType()
++ ~~~~~~~~~~~~~~~~
+    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
+    + FullyQualifiedErrorId : InvokeMethodOnNull
+ 
+True     True     ScriptBlock                              System.Object                                          
+True     True     Object[]                                 System.Array  
+#>
```
Comment (imba-tjd) [inline powershell.html.markdown:187]: same here

**Comment by AndrewDavis1191** [comment]
> ```
> function test ($value) { $value.gettype() }
> test []
> ```
> 
> The result is string
That's interesting, a null-valued expression seems to be converted to a string when passed through a function.
> 
> And I don't know whether you keep L172-L187 intentionally.
I did, but it seems unnecessary I suppose.
> 
> BTW I will need some days to review this because my mom only allows me to use pc for a limited time and I'm also a beginner of PS.
No worries, take your time.

File: powershell.html.markdown [hunk 6]
```diff
@@ -133,89 +373,54 @@ switch($val) {
...
 
-### Functions
-# The [string] attribute is optional.
-# Function names should follow Verb-Noun convention
-function Get-Foo([string]$name) {
-	echo "Hey $name, have a function"
+# Calling functions with parameters
+function Add-ParamNumbers {
+ param( [int]$FirstNumber, [int]$SecondNumber )
+ $FirstNumber + $SecondNumber
```
Comment (imba-tjd) [inline powershell.html.markdown:420]: How about using `Add-ParamNumbers([int]$FirstNumber, [int]$SecondNumber)`, which would make it more similar to other language.

Comment (AndrewDavis1191) [reply powershell.html.markdown:420]: There are later examples that illustrate this. I think in production you'd just go with whatever your team does.

**Comment by imba-tjd** [comment]
I have finished reading. My thoughts:

1. covering the usage of `&"xxx.exe"`. In any other shell language, `"xxx.exe"` is enough to call an external program, but for PS it only returns the original string.
2. When it refers to .net classes, the `System` can be omitted, e.g. `[System.Math]` -> `[Math]`
3. In `Show-Command Get-EventLog`, the `Get-EventLog` isn't exist in pwsh7.
4. PS doesn't support raw(bytes) piping: `curl.exe https://www.microsoft.com/favicon.ico -O` works, but `curl.exe https://www.microsoft.com/favicon.ico > favicon.ico` will break the image. It's worth mentioning this pitfall.

**Comment by AndrewDavis1191** [comment]
> 1. covering the usage of `&"xxx.exe"`. In any other shell language, `"xxx.exe"` is enough to call an external program, but for PS it only returns the original string.

I think in most cases, just typing the name works fine, but I see what you are referring to. If this will be included, then I think we should include dot sourcing as well for info on scoping the external commands run in this fashion.

> 2. When it refers to .net classes, the `System` can be omitted, e.g. `[System.Math]` -> `[Math]`

True, but I think in a tutorial we should just mention that in a comment. In the cases like [System.Collections.ArrayList] I think it is informative to see the drilldown in that way.

> 3. In `Show-Command Get-EventLog`, the `Get-EventLog` doesn't exist in pwsh7.

Ah yes, looks like it's deprecated in favor of "Get-WinEvent" [Issue #9852 on Powershell repo](https://github.com/PowerShell/PowerShell/issues/9852)

> 4. PS doesn't support raw(bytes) piping: `curl.exe https://www.microsoft.com/favicon.ico -O` works, but `curl.exe https://www.microsoft.com/favicon.ico > favicon.ico` will break the image. It's worth mentioning this pitfall.

Good idea, I don't think I've ever run into this issue personally; didn't even realize it was an issue.
I see there is and has been an open issue on this for quite some time. [Issue #1908 on Powershell repo](https://github.com/PowerShell/PowerShell/issues/1908)

> 5. I prefer keeping `1..10 | % { "Loop number $_" }` and some of the `| ?` because I think it's important to mention the pipeline idea.

I agree somewhat. Typically the pipeline is slower than alternative routes, but the convenience is invaluable. I would not use the % or ? or $_ aliases and instead recommend using the full command. It is more readable in my opinion.

**Comment by AndrewDavis1191** [comment]
Ah I see. The CI is failing currently due to a lint on [vim/de](https://github.com/adambard/learnxinyminutes-docs/pull/3999)

File: powershell.html.markdown [hunk 7]
```diff
@@ -554,6 +556,16 @@ True     False    Guitar                                   Instrument
 #>
 Get-Process | Foreach-Object ProcessName | Group-Object
 
+# Useful pipeline examples are iteration and filtering
+1..10 | ForEach-Object { "Loop number $PSITEM" }
+1..10 | where {$PSITEM -gt 5} | Format-Table
+
+# A noteable pitfall of the pipeline is it's performance when
+# compared with other options
+# additionally, raw bytes are not passed through the piipeline
```
Comment (imba-tjd) [inline powershell.html.markdown:565]: typo: piipeline

Comment (AndrewDavis1191) [reply powershell.html.markdown:565]: Gotta love the Drop mech keyboard quality.

**Comment by imba-tjd** [comment]
The point 1, my idea is that when the path contains spaces, in cmd you just need to add quotation marks:

```
> C:\Program Files\dotnet\dotnet.exe
'C:\Program' is not recognized as an internal or external command,
operable program or batch file.

> "C:\Program Files\dotnet\dotnet.exe"
...
(works)
```

But this experience doesn't work in PS:

```
> C:\Program Files\dotnet\dotnet.exe
C:\Program: The term 'C:\Program' is not recognized as a name of a cmdlet, function, script file, or executable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again

> "C:\Program Files\dotnet\dotnet.exe"
C:\Program Files\dotnet\dotnet.exe    # returns the string rather than execute it
```

Besides the arguments shouldn't be included in quotation marks:

```
&"C:\Program Files\dotnet\dotnet.exe --help"   # fail
&"C:\Program Files\dotnet\dotnet.exe" --help   # success
```

I think this is a common case when writing scripts.

Besides, I now find another thing worthy to mention. PS is using backticks to escape characters, which is different from most of the other languages. The benefit is that you needn't use `\\` in strings which is the directory seperator in Windows.

**Comment by AndrewDavis1191** [comment]
> The point 1, my idea is that when the path contains spaces, in cmd you just need to add quotation marks:

Ah, I see what you mean. I can add info on calling external exe's as well with the spaces

> ```
> &"C:\Program Files\dotnet\dotnet.exe --help"   # fail
> &"C:\Program Files\dotnet\dotnet.exe" --help   # success
> ```

For this purpose we can also mention dot-sourcing as an option
```
."C:\Program Files\dotnet\dotnet.exe" --help
```

> Besides, I now find another thing worthy to mention. PS is using backticks to escape characters, which is different from most of the other languages. The benefit is that you needn't (and couldn't) use `\\` in strings which is the directory seperator in Windows.

I intentionally left out any mention of backticks, since I just think they are ugly, and there are many workarounds for their uses that are much nicer. The escape chars have no alternative afaik though, so I'll include info on those.

File: powershell.html.markdown [hunk 8]
```diff
@@ -149,6 +149,18 @@ $age = 22
...
+# Escape Characters in Powershell
+# Many languages use the '\', but Windows uses this character for 
+# file paths. Powershell thus uses '`' to escape characters
+# Take caution when working with files, as '`' is a
+# valid character in NTFS filenames.
+"Showing`nEscape Chars" # => new line between Showing and Escape
+"Making`tTables`tWith`tTabs" # => Format things with tabs
+
+# Negate pound sign to prevent comment
+# Note that the function of '#' is removed, but '#" is still present
```
Comment (imba-tjd) [inline powershell.html.markdown:161]: Could you elaborate this line?

File: powershell.html.markdown [hunk 9]
```diff
@@ -13,116 +14,368 @@ Nearly all examples below can be a part of a shell script or executed directly
...
+2 -lt 3 -and 3 -lt 2  # => False
+
+# (-is vs. -eq) -is checks if two objects are the same type
+# -eq checks if the objects have the same values.
+# Note: we called '[Math]' from .NET previously without the preceeding
+# namespaces. We can do the same with [Collections.ArrayList] if preferred
+[System.Collections.ArrayList]$a = @()  # Point a at a new list
+$a = (1,2,3,4)
+$b = $a                                 # => Point b at what a is pointing to
+$b -is $a.getType()                     # => True, a and b equal same type
```
Comment (imba-tjd) [inline powershell.html.markdown:125]:
```
GetType
```

**Review by imba-tjd — APPROVED** [review]
🆗 That's all my reviews.

**Comment by AndrewDavis1191** [comment]
@imba-tjd I really appreciate you taking the time to review this! Thank you! :Grin:

**Pull request merged**